coverage/sqldata.py

```diff
@@ -146,22 +153,23 @@ def _open_db(self):
                         )
                     )
 
-            for row in self._db.execute("select has_lines, has_arcs from meta"):
+            for row in db.execute("select has_lines, has_arcs from meta"):
                 self._has_lines, self._has_arcs = row
 
-            for path, id in self._db.execute("select path, id from file"):
+            for path, id in db.execute("select path, id from file"):
                 self._file_map[path] = id
 
     def _connect(self):
-        if self._db is None:
+        if get_thread_id() not in self._dbs:
             if os.path.exists(self.filename):
```

Comment:
@strichter I'll merge this today, but I'm wondering about a race condition on the creation of the database.  There's a window here between checking if the file exists, and writing the initial metadata into the database.  Is there a reason we don't see failures due to that race?

Reply:
Good call! I think that coverage starts collecting so early that it will always hit the initialization case before any threads are spawned. This is at least true in our setup where we run coverage as a shell script. In other words the first time this condition hits, we are in the main thread.